flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/WindowOperatorBuilder.java

```diff
@@ -181,24 +194,79 @@ public WindowOperatorBuilder aggregate(
 			LogicalType[] accumulatorTypes,
 			LogicalType[] aggResultTypes,
 			LogicalType[] windowPropertyTypes) {
-		this.accumulatorTypes = accumulatorTypes;
-		this.aggResultTypes = aggResultTypes;
-		this.windowPropertyTypes = windowPropertyTypes;
+		aggregate(accumulatorTypes, aggResultTypes, windowPropertyTypes);
 		this.generatedAggregateFunction = checkNotNull(generatedAggregateFunction);
 		this.generatedEqualiser = checkNotNull(generatedEqualiser);
 		return this;
 	}
 
+	public WindowOperatorBuilder aggregate(
+			NamespaceTableAggsHandleFunction<?> tableAggregateFunction,
+			LogicalType[] accumulatorTypes,
+			LogicalType[] aggResultTypes,
+			LogicalType[] windowPropertyTypes) {
+		ClosureCleaner.clean(tableAggregateFunction, ExecutionConfig.ClosureCleanerLevel.RECURSIVE, true);
+		aggregate(accumulatorTypes, aggResultTypes, windowPropertyTypes);
+		this.tableAggregateFunction = checkNotNull(tableAggregateFunction);
+		return this;
+	}
+
+	public WindowOperatorBuilder aggregate(
+			GeneratedNamespaceTableAggsHandleFunction<?> generatedTableAggregateFunction,
+			LogicalType[] accumulatorTypes,
+			LogicalType[] aggResultTypes,
+			LogicalType[] windowPropertyTypes) {
+		aggregate(accumulatorTypes, aggResultTypes, windowPropertyTypes);
+		this.generatedTableAggregateFunction = checkNotNull(generatedTableAggregateFunction);
+		return this;
+	}
+
+	@VisibleForTesting
+	public WindowOperatorBuilder aggregate(
+		NamespaceAggsHandleFunctionBase<?> aggregateFunction,
+		RecordEqualiser equaliser,
+		LogicalType[] accumulatorTypes,
+		LogicalType[] aggResultTypes,
+		LogicalType[] windowPropertyTypes) {
+		if (aggregateFunction instanceof NamespaceAggsHandleFunction) {
+			aggregate((NamespaceAggsHandleFunction) aggregateFunction,
+				equaliser,
+				accumulatorTypes,
+				aggResultTypes,
+				windowPropertyTypes);
+		} else {
+			aggregate((NamespaceTableAggsHandleFunction) aggregateFunction,
+				accumulatorTypes,
+				aggResultTypes,
+				windowPropertyTypes);
+		}
+		return this;
+	}
+
 	public WindowOperatorBuilder withSendRetraction() {
 		this.sendRetraction = true;
 		return this;
 	}
 
 	public WindowOperator build() {
 		checkNotNull(trigger, "trigger is not set");
-		if (generatedAggregateFunction != null && generatedEqualiser != null) {
+		if (generatedTableAggregateFunction != null) {
```

Comment:
There are a lot of branches, can we abstract this builder to multiple builders?

Reply:
Good idea. We can make the WindowOperatorBuilder as a base builer and add two more builders for aggregate and table aggregate.
What do you think?